src/custom-apps/components/PermissionAlert/utils.ts

```diff
@@ -0,0 +1,145 @@
+  fields: Array<{
+    name: string;
+    description: string;
+  }>;
+}
+
+interface PermissionChildNode {
+  permissions: string[];
+  children: string[];
+}
+
+interface PermissionNode {
+  permissions: string[];
+  children: Record<string, PermissionChildNode>;
+}
+
+type PermissionMap = Record<string, PermissionNode>;
+
+// cannot be in `queries.ts` as codegen cannot handle `__schema`
+export const IntrospectionQuery = gql`
+  query PermissionIntrospection {
+    __schema {
+      types {
+        kind
+        name
+        description
+        fields(includeDeprecated: false) {
+          name
+          description
+        }
+      }
+    }
+  }
+`;
+
+const uniq = <T>(value: T, index: number, self: T[]) =>
+  self.indexOf(value) === index;
+
+const extractPermissions = (description?: string) => {
+  const match = (description || "").match(/following permissions(.*): (.*?)\./);
+  const permissions = match ? match[2].split(",") : [];
+
+  return permissions;
+};
+
+export const getPermissions = (
+  query: string,
+  permissionMapping: PermissionMap,
+) => {
+  const cursors = extractCursorsFromQuery(query);
+  return cursors.map(findPermission(permissionMapping)).flat().filter(uniq);
+};
+
+export const buildPermissionMap = (
+  elements: IntrospectionNode[],
+): PermissionMap =>
+  elements
+    .filter(({ kind }) => kind === "OBJECT")
+    .filter(
+      ({ name }) =>
+        !/(Created|Create|Delete|Deleted|Update|Updated)$/.test(name),
+    )
+    .reduce((saved, { name, description, fields }) => {
+      const permissions = extractPermissions(description);
+
+      const children = fields.reduce((prev, { name, description }) => {
+        const permissions = extractPermissions(description);
+
+        return {
+          ...prev,
+          [name.toLowerCase()]: { permissions },
+        };
+      }, {});
+
+      return {
+        ...saved,
+        [name.toLowerCase()]: {
+          permissions,
+          children,
+        },
+      };
+    }, {});
+
+const byKind = (name: string) => (element: SelectionNode) =>
+  element.kind === name;
+const extractValue = (element: FieldNode) => element.name.value;
+const isNotEvent = (value: string) => value !== "event";
+
+export const extractCursorsFromQuery = (query: string) => {
+  const cursors = [] as string[][];
+
+  try {
+    const ast = parse(query);
+
+    visit(ast, {
+      Field(node, _key, _parent, _path, ancestors) {
+        if (node.name.value !== "__typename") {
+          const cursor = ancestors
+            .filter(byKind("Field"))
+            .map(extractValue)
+            .filter(isNotEvent);
+
+          if (cursor.length > 0) {
+            cursors.push([...cursor, node.name.value]);
+          }
+        }
+      },
+    });
+  } catch (error) {
+    // explicit silent
+  }
+
+  return cursors;
+};
+
+const groupBy = <T>(list: T[], groupSize = 2) =>
+  list
+    .slice(groupSize - 1)
+    .map((_, index) => list.slice(index, index + groupSize));
```

Comment:
**suggestion**: It's a little unclear what it does. `groupBy` indicates we're grouping by some kind of iteratee, but we can only provide group size. Also - resulting array consists of *overlapping* subarrays, which is not so obvious without diving into logic. I'd appreciate explicit return type here.

Reply:
@Droniu added a comment that clarifies that part